### ISSUE TITLE
S3900 FP: with string.IsNullOrEmpty

### ISSUE THREAD

**Issue by CFlatWouldbeMinor** [SonarSource/sonar-dotnet#2670]
### Description

This might be related to #1347, but with new Analyzer 7-16-0-8981 I am geting a validation warning S3900 when validating with IsNullOrEmpty

### Repro steps

        public static string MyFunc(string argument)
        {
            if (string.IsNullOrEmpty(argument) == true)
            {
                return null;
            }
            if (SemanticTagHelper.AllRegexPattern.IsMatch(argument) == true)
            {
                int index = argument.LastIndexOf

### Expected behavior

In my opinion IsNullOrEmpty should be sufficient for validation

### Actual behavior

In Visual studio this is reported as warning

* SonarC# Version : 7-16-0-8981
* Visual Studio Version VS2019 16.3
* If running through the Scanner for MSBuild, its version

**Comment by andrei-epure-sonarsource** [comment]
thanks @CFlatWouldbeMinor , I confirm the FP

What's interesting is that if you don't do the equality comparison to true, the FP disappears
```
        public static void NoComparison(string argument, bool b)
        {
            if (string.IsNullOrEmpty(argument))
            {
                return;
            }
            if (b)
            {
                int index = argument.LastIndexOf('c'); // no issue here, all good
            }
        }
```

**Comment by csaba-sagi-sonarsource** [comment]
I added string.IsNullOrWhitespace() case to the repros.

**Comment by StingyJack** [comment]
> What's interesting is that if you don't do the equality comparison to `true`, the FP disappears

@csaba-sagi-sonarsource - also note that removing the equality comparison to `true` only resolves the usage immediately following that check. usages after that check are still flagged as S3900.  This is a function we have that complains with S3900 about both `valueToCheck` and `invalidChars`

``` csharp
        public static bool ValidateParameter(ref string valueToCheck, bool checkForNull, bool checkIfEmpty, char[] invalidChars, int maxSize)
        {
            if (string.IsNullOrWhiteSpace(valueToCheck)) //originally this was if (valueToCheck == null), so I changed it to string.IsNullOr.... == true), and then to whats here now.
            {
                return !checkForNull;
            }

            valueToCheck = valueToCheck.Trim(); //<-- this used to be an S3900 complaint but is not singe I changed the first statement in this method

            var valid = (checkIfEmpty == false || (valueToCheck.Length >= 1)) //<-- S3900 complaint
                        && ((maxSize <= 0) || (valueToCheck.Length <= maxSize)); //<-- S3900 complaint
            
            if (valid && invalidChars.Length > 0)  //<-- S3900 complaint
            {
                var chars = invalidChars; //<-- no complaint
                var i = 0; 
                while (valid && i < chars.Length)
                {
                    valid &= !valueToCheck.Contains(chars[i]); //<-- S3900 complaint
                    i++;
                }
            }

            return valid;
        }
```

I'll add a param null check for the `char[]` for that one, but I thought it was interesting how the removal of the `== true` would have any effect and that it would change only part of it.   Its important that rule S3900 not assume that any of the other rules are configured and in effect (specifically the errant redundant boolean literal one).

**Comment by pavel-mikula-sonarsource** [comment]
Fixed by #6793
